feat(tree): Add Cdk Tree component

tinayuangao opened the pull request:
An initial version of cdk-tree

For #3175

jelbourn commented on src/cdk/collections/tree-adapter.ts:
```diff
@@ -0,0 +1,18 @@
+
+  expandFlattenedNodes(nodes: T[], expansionModel: SelectionModel<T>): T[];
+
+  nodeDecedents(node: T, nodes: T[], onlyExpandable: boolean);
```
Add JsDoc description for these methods?

jelbourn commented on src/cdk/tree/nested-node.ts:
```diff
@@ -0,0 +1,85 @@
+
+
+/**
+ * Nested node, add children to `cdkNodePlaceholder` in template.
```
Clarify that this is a child of `<cdk-tree>` and what is means to be a "nested node"? I'm also not sure exactly what ``"add children to `cdkNodePlaceholder` in template."`` means

jelbourn commented on src/cdk/tree/nested-node.ts:
```diff
@@ -0,0 +1,85 @@
+  protected _childrenSubscription: Subscription;
+
+  /** Subscribes to node placeholder changes. */
+  protected _placeholderSubscription: Subscription;
```
One pattern we've been using more lately is to create a subject like `_destroyed` and then use `takeUntil(this._destroyed)` to avoid needing to manually keep track of the subscription and unsubscribe

jelbourn commented on src/cdk/tree/nested-node.ts:
```diff
@@ -0,0 +1,85 @@
+  @ContentChildren(CdkNodePlaceholder) nodePlaceholder: QueryList<CdkNodePlaceholder>;
+
+  /** The Children nodes data. */
+  protected _children: T[];
```
I'm not very clear on what this represents exactly

jelbourn commented on src/cdk/tree/node.ts:
```diff
@@ -0,0 +1,112 @@
+import {FlatNode, NestedNode} from './tree-data';
+
+/** The tree node template */
+export const CDK_TREE_NODE_TEMPLATE = '<ng-content cdkNodeOutlet></ng-content>';
```
Any reason not to just put this inline in the component metadata?

jelbourn commented on src/cdk/tree/nested-node.ts:
```diff
@@ -0,0 +1,85 @@
+@Directive({
+  selector: '[cdkNestedNode]'
+})
+export class CdkNestedNode<T extends NestedNode> implements AfterContentInit, OnDestroy {
```
Should this be `CdkNestedTreeNode`? I'm not totally clear on the different between this and `CdkTreeNode`

jelbourn commented on src/cdk/tree/node.ts:
```diff
@@ -0,0 +1,112 @@
+   */
+  when: (nodeData: T, index: number) => boolean;
+
+  constructor(public template: TemplateRef<any>) {}
```
Should `template` be `@docs-private`?

jelbourn commented on src/cdk/tree/node.ts:
```diff
@@ -0,0 +1,112 @@
+/**
+ * Tree node for CdkTree. It contains the data in the tree node.
+ */
+// TODO: Role should be group for expandable ndoes
```
Can you move the `TODO` next to where `role` is defined? When there's another comment between the JsDoc and the class definition dgeni doesn't connect them

jelbourn commented on src/cdk/tree/node.ts:
```diff
@@ -0,0 +1,112 @@
+    if (this._childrenSubscrition) {
+      this._childrenSubscrition.unsubscribe();
+    }
+    if ('level' in v) {
```
We can't use `in` with string-based property names since Closure Compiler will rename properties

jelbourn commented on src/cdk/tree/node.ts:
```diff
@@ -0,0 +1,112 @@
+      this.role = (this._data as FlatNode).expandable ? 'group' : 'treeitem';
+    } else {
+      this._childrenSubscrition = (this._data as NestedNode).getChildren().subscribe((children) => {
+        this.role = !!children ? 'group' : 'treeitem';
```
Don't need to use `!!` in conditionals, only when returning / storing the value

jelbourn commented on src/cdk/tree/node.ts:
```diff
@@ -0,0 +1,112 @@
+
+
+  /** The offset top of the element. Used by CdkTree to decide the order of the nodes. [Focus] */
+  get offsetTop() {
```
Should this be internal?

jelbourn commented on src/cdk/tree/node.ts:
```diff
@@ -0,0 +1,112 @@
+   * The role of the node should be 'group' if it's an internal node,
+   * and 'treeitem' if it's a leaf node.
+   */
+  @Input() role: string = 'treeitem';
```
Instead of `string`, should this be `'treeitem' | 'group'`?

jelbourn commented on src/cdk/tree/node.ts:
```diff
@@ -0,0 +1,112 @@
+    }
+  }
+
+  /** Focuses the menu item. Implements for FocusableOptoin. */
```
`FocusableOptoin` -> `FocusableOption`

jelbourn commented on src/cdk/tree/padding.ts:
```diff
@@ -0,0 +1,40 @@
+@Directive({
+  selector: '[cdkNodePadding]',
+  host: {
+    '[style.padding-left]': 'paddingIndent',
```
Need to account for RTL?

jelbourn commented on src/cdk/tree/padding.ts:
```diff
@@ -0,0 +1,40 @@
+  @Input('cdkNodePadding') level: number;
+
+  /** The indent for each level. Default value is 28px. */
+  @Input('cdkNodePaddingIndex') indent: number = 28;
```
Any background on where `28px` came from?

jelbourn replied:
This is marked as an `Input` but it has no setter. Also is `Index` the right term?

jelbourn commented on src/cdk/tree/padding.ts:
```diff
@@ -0,0 +1,40 @@
+  /** The indent for each level. Default value is 28px. */
+  @Input('cdkNodePaddingIndex') indent: number = 28;
+
+  get paddingIndent() {
```
Can this be a method instead of a getter?

jelbourn commented on src/cdk/tree/placeholder.ts:
```diff
@@ -0,0 +1,23 @@
+@Directive({
+  selector: '[cdkNodePlaceholder]'
+})
+export class CdkNodePlaceholder {
```
`Placeholder` -> `Outlet`?

jelbourn commented on src/cdk/tree/select-trigger.ts:
```diff
@@ -0,0 +1,50 @@
+  constructor(@Inject(forwardRef(() => CdkTree)) private tree: CdkTree<T>,
+              private treeNode: CdkTreeNode<T>) {}
+
+  trigger(_: Event) {
```
Can you not omit the event argument?

jelbourn commented on src/cdk/tree/select-trigger.ts:
```diff
@@ -0,0 +1,50 @@
+    '(click)': 'trigger($event)',
+  }
+})
+export class CdkNodeSelectTrigger<T> {
```
I don't quite remember why this is baked into the tree- what's the background here?

tinayuangao replied:
This can be used to change expand/collapse a tree node/subtree with a custom SelectionModel(a more general trigger for CdkNodeTrigger). Removed since it should not be included for the initial version.

jelbourn commented on src/cdk/tree/tree-control.ts:
```diff
@@ -0,0 +1,223 @@
+import {CdkTreeNode} from './node';
+
+/**
+ * Tree control interface
```
Can you expand this comment a bit to mention how you would use this interface?

jelbourn commented on src/cdk/tree/tree-control.ts:
```diff
@@ -0,0 +1,223 @@
+/**
+ * Tree control interface
+ */
+export interface TreeControl {
```
Should this be generic to avoid using `any`?

jelbourn commented on src/cdk/tree/tree-control.ts:
```diff
@@ -0,0 +1,223 @@
+  expanded(node: any): boolean;
+
+  /** Get all decedents of a node */
+  getDecedents(node: any): any[];
```
`Decedents` -> `Descendants` here and elsewhere

jelbourn commented on src/cdk/tree/tree-control.ts:
```diff
@@ -0,0 +1,223 @@
+}
+
+/** Base tree control. It has basic toggle/expand/collapse operations on a single node. */
+export abstract class BaseTreeControl<T extends FlatNode|NestedNode> implements TreeControl {
```
What do you think of making a subdirectory for controls with separate files for `tree-control.ts`, `base-tree-control.ts`, `flat-tree-control.ts` and `nested-tree-control.ts`?

jelbourn commented on src/cdk/tree/tree-control.ts:
```diff
@@ -0,0 +1,223 @@
+
+  /** Expands a subtree rooted at given `node` recursively. */
+  expandDecedents(node: T) {
+    let decedents = this.getDecedents(node);
```
Prefer `const` (here and elsewhere)

jelbourn commented on src/cdk/tree/tree-control.ts:
```diff
@@ -0,0 +1,223 @@
+   */
+  expandAll() {
+    this.expansionModel.clear();
+    this.nodes.forEach((node) => this._expandDecedents(node));
```
Nit: don't need parens for a single argument with inferred type (here and elsewhere)

jelbourn commented on src/cdk/tree/tree-control.ts:
```diff
@@ -0,0 +1,223 @@
+      if (children) {
+        children.forEach((child) => {
+          this.collapseDecedents(child)
+        });
```
Don't need to use braces here
```ts
children.forEach(child => this.collapseDecedents(child));
```
(similar elsewhere)

jelbourn commented on src/cdk/tree/tree-data.ts:
```diff
@@ -0,0 +1,20 @@
+
+/** Interface for data in flat tree. Tree users should implement this interface to use CdkTree. */
+export interface FlatNode {
+  level: number;
```
Should we do a `getLevel` method instead of a property? It might be annoying if someone has a `level` property in their data that means something else. (Same for `expandable` -> `isExpandable`). 

I also wonder if there's a way to could express this in the template (similar to `when`) instead of requiring an interface, but I forget if we already explored that option

tinayuangao replied:
We didn't explore that option. 

I can try to express it in the template, but `TreeControl` is also using that to expand/collapse subtrees/the whole tree.

If we want to express this in the template, we can also pass that info to `TreeControl`, or put `getLevel` `getChildren` and `isExpandable` functions in `TreeControl` and let other components use functions in `TreeControl`.

tinayuangao replied:
Now moved `getLevel` `isExpandable` and `getChildren` to `TreeControl`. Removed `NestedNode` and `FlatNode` types and users can use any type of data.

jelbourn commented on src/cdk/tree/tree.spec.ts:
```diff
@@ -0,0 +1,161 @@
+        SimpleCdkTreeApp,
+        // DynamicDataSourceCdkTreeApp,
+        // NodeContextCdkTreeApp,
+        // WhenNodeCdkTreeApp
```
Add `TODO: ` in front of each of these? That will change them from commented-out-code to just comments

tinayuangao commented:
Comments addressed. Please take a look. Thanks!

jelbourn commented on src/cdk/collections/tree-adapter.ts:
```diff
@@ -0,0 +1,28 @@
+ */
+export interface TreeAdapter<T> {
+  /** Flatten structured data to an array of data. */
+  flattenNodes( structuredData: any[]): T[];
```
Extra space

jelbourn commented on src/cdk/collections/tree-adapter.ts:
```diff
@@ -0,0 +1,28 @@
+   * Put node descendents of node in array.
+   * If `onlyExpandable` is true, then only process expandable descendents.
+   */
+  nodeDecedents(node: T, nodes: T[], onlyExpandable: boolean);
```
"Decedents" -> "Descendants"
(here and elsewhere)

jelbourn commented on src/cdk/collections/tree-adapter.ts:
```diff
@@ -0,0 +1,28 @@
+/**
+ * @license
+ * Copyright Google LLC. All Rights Reserved.
```
Extra `.` after LLC

jelbourn commented on src/cdk/tree/control/base-tree-control.ts:
```diff
@@ -0,0 +1,80 @@
+  nodes: T[];
+
+  /** Expansion info: the changes */
+  expandChange = new BehaviorSubject<T[]>([]);
```
Maybe `expansionChange`? @andrewseguin thoughts?

jelbourn commented on src/cdk/tree/control/base-tree-control.ts:
```diff
@@ -0,0 +1,80 @@
+  }
+
+  /** Whether a given node is expanded or not. Returns true if the node is expanded. */
+  expanded(node: T) {
```
`isExpanded`?

jelbourn commented on src/cdk/tree/control/base-tree-control.ts:
```diff
@@ -0,0 +1,80 @@
+  abstract expandAll();
+
+  /** Expands a subtree rooted at given `node` recursively. */
+  abstract expandDescendents(node: T);
```
"Descendents" -> "Descendants"
(here and elsewhere)

jelbourn commented on src/cdk/tree/control/base-tree-control.ts:
```diff
@@ -0,0 +1,80 @@
+  /** Whether the node is expandable. Returns true if expandable. This is for flat tree node. */
+  isExpandable: (node: T) => boolean;
+
+  /** Returns an observable for a given node's children. This is for nested tree node. */
```
"Gets a stream that emits whenever the given node's children change"

Is that accurate?

jelbourn commented on src/cdk/tree/control/tree-control.ts:
```diff
@@ -0,0 +1,65 @@
+ * The CDKTree will use this TreeControl to expand/collapse a node.
+ * User can also use it outside the `<cdk-tree>` to control the expansion status of the tree.
+ */
+export interface TreeControl<T> {
```
Unit tests for each type of TreeControl?

jelbourn commented on src/cdk/tree/node.ts:
```diff
@@ -0,0 +1,112 @@
+  /** The tree node data */
+  @Input('cdkTreeNode')
+  set data(v: T) {
+    this._data = v;
```
Break the content of this setter into a separate function?

jelbourn commented on src/cdk/tree/node.ts:
```diff
@@ -0,0 +1,112 @@
+    }
+  }
+  get data(): T { return this._data; }
+  _data: T;
```
Make this private?

jelbourn commented on src/cdk/tree/node.ts:
```diff
@@ -0,0 +1,112 @@
+  _data: T;
+
+  /** The offset top of the element. Used by CdkTree to decide the order of the nodes. [Focus] */
+  private get offsetTop() {
```
This this is `private`, make it a method?

jelbourn commented on src/cdk/tree/padding.ts:
```diff
@@ -0,0 +1,54 @@
+    return this._dir && this._dir.value === 'rtl' ? this._paddingIndent() : null;
+  }
+
+  constructor(private _treeNode: CdkTreeNode<T>,
```
Move constructor above methods and below properties

jelbourn commented on src/cdk/tree/padding.ts:
```diff
@@ -0,0 +1,54 @@
+
+  constructor(private _treeNode: CdkTreeNode<T>,
+              private _tree: CdkTree<T>,
+              @Optional() private _dir: Directionality) {}
```
Probably want to have `dir.change` as well

jelbourn commented on src/cdk/tree/trigger.ts:
```diff
@@ -0,0 +1,39 @@
+})
+export class CdkNodeTrigger<T> {
+  /** Whether expand/collapse the node recursively. */
+  @Input('cdkNodeTriggerRecursive') recursive: boolean = true;
```
`coerceBooleanProperty`?

jelbourn commented on src/cdk/tree/trigger.ts:
```diff
@@ -0,0 +1,39 @@
+ * Node trigger to expand/collapse the node.
+ */
+@Directive({
+  selector: '[cdkNodeTrigger]',
```
`cdkNodeTrigger` doesn't seem quite specific enough. Maybe `cdkTreeNodeToggle`?

jelbourn commented on src/cdk/tree/tree.spec.ts:
```diff
@@ -0,0 +1,139 @@
+import {CdkTree} from './tree';
+
+
+describe('CdkTree', () => {
```
In a follow-up PR, it would be good to explore creating a syntax-sugar way of testing tree content. The table, for example, has this:
```ts
      expectTableToMatchContent(tableElement, [
        ['Column A', 'Column B', 'Column C'],
        ['a_1', 'b_1', 'c_1'],
        ['a_2', 'b_2', 'c_2'],
        ['a_3', 'b_3', 'c_3'],
      ]);
```

For the tree, we'd need to capture level and expansion state

jelbourn commented on src/cdk/tree/tree-module.ts:
```diff
@@ -0,0 +1,35 @@
+import {CdkNodeDef, CdkTreeNode} from './node';
+import {CdkNestedTreeNode} from './nested-node';
+
+const EXPORTED_DECLARATIONS = [
```
Maybe just call this `TREE_DIRECTIVES`?

crisbeto commented on src/cdk/tree/control/base-tree-control.ts:
```diff
@@ -0,0 +1,80 @@
+  expansionModel: SelectionModel<T> = new SelectionModel<T>(true);
+
+  /** Toggles one single node. Expands a collapsed node or collapse an expanded node. */
+  toggle(node: T) {
```
All of these public methods need return types. Same goes for the other files.

crisbeto commented on src/cdk/tree/control/base-tree-control.ts:
```diff
@@ -0,0 +1,80 @@
+  /** Expands one single node. */
+  expand(node: T) {
+    this.expansionModel.select(node);
+    this.expansionChange.next(this.expansionModel.selected);
```
You should be able to replace the `expansionChange` event with the `onChange` from the `SelectionModel`.

crisbeto commented on src/cdk/tree/control/nested-tree-control.ts:
```diff
@@ -0,0 +1,79 @@
+   */
+  expandAll() {
+    this.expansionModel.clear();
+    this.nodes.forEach((node) => this._expandDescendants(node));
```
nit: you don't need the parentheses around `node`.

crisbeto commented on src/cdk/tree/control/nested-tree-control.ts:
```diff
@@ -0,0 +1,79 @@
+        children.forEach((child: T) => this._getDescendants(descendants, child));
+      }
+    });
+    subscription.unsubscribe();
```
Rather than unsubscribing manually here, you could use the `first` operators instead.

crisbeto commented on src/cdk/tree/node.ts:
```diff
@@ -0,0 +1,116 @@
+      if (!this._tree.treeControl.getChildren) {
+        throw getTreeControlFunctionsMissingError();
+      }
+      takeUntil.call(this._tree.treeControl.getChildren(this._data), this._destroyed)
```
Since this subscription is set up by the setter, shouldn't it unsubscribe from any previous subscriptions? Otherwise you could end up with multiple subscriptions if the `data` is set multiple times.

crisbeto commented on src/cdk/tree/node.ts:
```diff
@@ -0,0 +1,116 @@
+  }
+
+  /** The offset top of the element. Used by CdkTree to decide the order of the nodes. [Focus] */
+  private getOffsetTop() {
```
I don't see this method being used anywhere, is it still relevant?

tinayuangao replied:
Removed.

crisbeto commented on src/cdk/tree/tree.ts:
```diff
@@ -0,0 +1,272 @@
+
+  ngAfterViewInit() {
+    // For key traversal in correct order
+    this.items.changes.subscribe(items => {
```
It seems like you're never unsubscribing from this one and I'm not sure whether Angular will complete the stream on destroy.

crisbeto commented on src/cdk/tree/tree.ts:
```diff
@@ -0,0 +1,272 @@
+      this.orderedNodes.reset(nodes);
+
+      const activeItem = this._keyManager ? this._keyManager.activeItem : null;
+      this._keyManager = new FocusKeyManager(this.orderedNodes);
```
Since you're mutating the same `QueryList`, you shouldn't need to re-create the key manager every time the items change. You may have to reset the active index though.

tinayuangao replied:
Reset it because the order of the items in the `QueryList` is wrong. For example:
- node 1
- node 2
After expanded node 1, it becomes 
- node 1
---- node 3
---- node 4
-node 2
in item list, and UP_ARROW and DOWN_ARROW will jump from one node to another.
Resetting the nodes make it become
- node 1
---- node 2
---- node 3
- node 4
in key manager.

tinayuangao replied:
Removed key manager related part.

crisbeto commented on src/cdk/tree/tree.ts:
```diff
@@ -0,0 +1,272 @@
+    if (event.keyCode == UP_ARROW) {
+      this._keyManager.setPreviousItemActive();
+    } else if (event.keyCode == DOWN_ARROW) {
+      this._keyManager.setNextItemActive();
```
The UP_ARROW and DOWN_ARROW behavior should be handled by the key manager already.

tinayuangao commented:
- Added tests for flat-tree-control and nested-tree-control.
- Removed `cdkTreeNode` input from `CdkTreeNode` and put the data through static variable. 
- Removed key manager related part.

Please take a look. Thanks for review!

crisbeto commented on src/cdk/tree/nested-node.ts:
```diff
@@ -0,0 +1,91 @@
+        this._addChildrenNodes();
+      });
+    takeUntil.call(this.nodeOutlet.changes, this._destroyed)
+      .subscribe((_) => this._addChildrenNodes());
```
The `_` param here is unnecessary. We may have a lint rule for unused parameters that would complain about this.

andrewseguin replied:
Looks like this is still here

crisbeto commented on src/cdk/tree/node.ts:
```diff
@@ -0,0 +1,123 @@
+   */
+  when: (nodeData: T, index: number) => boolean;
+
+  /** @docs-private */
```
You should be able to set this on the directive doc, rather than the constructor.

crisbeto commented on src/cdk/tree/padding.ts:
```diff
@@ -0,0 +1,54 @@
+  }
+
+  /** The left padding indent value for the tree node. */
+  paddingIndentLeft(): string|null {
```
Rather than having two properties and bindings for left and right, you could set the `padding-left`/`padding-right` instead. E.g.:

```
setPadding() {
  const padding = this._paddingIndent();
  const paddingProp = this._dir && this._dir.value === 'rtl' ? 'padding-right' : 'padding-left';
  
  this._renderer.setStyle(this._element.nativeElement, paddingProp, padding);
}
```

tinayuangao replied:
Updated `padding.ts`. Now set the properties and listen to `dir.change`.

jelbourn commented on src/cdk/collections/tree-adapter.ts:
```diff
@@ -0,0 +1,28 @@
+ * Tree Adapter to change any structured data type to flattened tree data.
+ * To be implemented by tree users.
+ */
+export interface TreeAdapter<T> {
```
I'm wondering if a more specific name might be useful here. Maybe something like `TreeFlattener` or `FlatTreeAdapter`. Thoughts?

tinayuangao replied:
`TreeDataNodeFlattener`?

jelbourn commented on src/cdk/tree/control/flat-tree-control.ts:
```diff
@@ -0,0 +1,43 @@
+  /**
+   * Expands all nodes in the tree.
+   *
+   * To make this working, the `nodes` of the TreeControl must be set correctly.
```
It's not clear what "the `nodes` of the TreeControl must be set correctly" means

jelbourn commented on src/cdk/tree/tree.spec.ts:
```diff
@@ -0,0 +1,201 @@
+});
+
+export class TestData {
+  a: string;
```
Make up fake properties for this? I'm a fan of _obviously fake_ testing data like `ponyName` or `pizzaTopping` since it  is clearly distinguished from actual component APIs

jelbourn commented on src/cdk/tree/tree.spec.ts:
```diff
@@ -0,0 +1,201 @@
+
+  constructor() {
+    super();
+    for (let i = 0; i < 3; i++) { this.addData(); }
```
`for` body should go on its own line

jelbourn commented on src/cdk/tree/tree.ts:
```diff
@@ -0,0 +1,216 @@
+  private _destroyed = new Subject<void>();
+
+  /** Latest data provided by the data source through the connect interface. */
+  private _data: Array<T> = [];
```
`T[]`

jelbourn commented on src/cdk/tree/node.ts:
```diff
@@ -0,0 +1,123 @@
+/** Context provided to the ndoes */
+export class CdkTreeNodeOutletContext<T> {
+
+  static mostRecentContextData: any;
```
This is slightly different from what I had in mind. My proposal had been that immediately after calling `createEmbeddedView`, you get the most recent node directive and set the data to it. I think this would be slightly less brittle, since its more reliable than depending on the order of instantiation of the context objects. 

For whatever we end up doing, we should have some very descriptive comments explaining the situation.

tinayuangao replied:
Now we use a static variable in `CdkTreeNode` and set data to that `mostRecentTreeNode`.

jelbourn commented on src/cdk/tree/tree.ts:
```diff
@@ -0,0 +1,216 @@
+  /** The tree controller */
+  @Input() treeControl: TreeControl<T>;
+
+  // TODO(andrewseguin): Remove max value as the end index
```
Change `TODO` to you?

jelbourn commented on src/cdk/tree/node.ts:
```diff
@@ -0,0 +1,123 @@
+import {getTreeControlFunctionsMissingError} from './tree-errors';
+
+
+/** Context provided to the ndoes */
```
ndoes

jelbourn commented on src/cdk/tree/padding.ts:
```diff
@@ -0,0 +1,74 @@
+    this._indent = value;
+    this._setPadding();
+  }
+  _indent: number = 40;
```
Should this support people setting a string with units? E.g. `10em`. 

(could be in a follow-up PR, but maybe add a TODO)

jelbourn commented on src/cdk/tree/padding.ts:
```diff
@@ -0,0 +1,74 @@
+  /** The level of depth of the tree node. The padding will be `level * indent` pixels. */
+  @Input('cdkTreeNodePadding')
+  set level(value: number) {
+    this._level = value;
```
`coerceNumberProperty`?

jelbourn commented on src/cdk/tree/nested-node.ts:
```diff
@@ -0,0 +1,91 @@
+ * this:
+ *   <cdk-tree-node cdkNestedTreeNode [cdkNode]="node">
+ *     tree node data: {{node.name}}
+ *     <ng-template cdkNodeOutlet>
```
Should remove the `ng-template` element from the output in this example

jelbourn commented on src/cdk/tree/nested-node.ts:
```diff
@@ -0,0 +1,91 @@
+@Directive({
+  selector: '[cdkNestedTreeNode]'
+})
+export class CdkNestedTreeNode<T> implements AfterContentInit, OnDestroy {
```
`CdkNestedTreeNode` isn't template-ized the same way that flat tree nodes are now. Does that limit us in the future if we want to change up the DOM structure of the nested tree?

tinayuangao replied:
Now `CdkNestedTreeNode` extends `CdkTreeNode`. Like `CdkTreeNode`, users can put anything in `<cdk-nested-tree-node>` as template, but the children nodes will be added to `[cdkTreeNodeOutlet]` inside the nested node template. If there's no `cdkTreeNodeOutlet`, it works the same way as `CdkTreeNode`.

jelbourn commented on src/cdk/tree/outlet.ts:
```diff
@@ -0,0 +1,22 @@
+ * inside the outlet.
+ */
+@Directive({
+  selector: '[treeNodeOutlet]'
```
`cdkTreeNodeOutlet`?

jelbourn commented on src/cdk/tree/tree.ts:
```diff
@@ -0,0 +1,216 @@
+  moduleId: module.id,
+  selector: 'cdk-tree',
+  exportAs: 'cdkTree',
+  template: `<ng-container treeNodeOutlet></ng-container>`,
```
Does this PR have the handling for nested tree nodes? It looks like they'd just be dropped as-is.

tinayuangao replied:
Yes. The nested node handling will be done in `CdkNestedTreeNode` directive.

tinayuangao commented:
@jelbourn take a look when you have time? Thanks!

jelbourn commented:
@tinayuangao up to you whether you want @andrewseguin to look before merging

andrewseguin commented on src/cdk/tree/control/base-tree-control.ts:
```diff
@@ -0,0 +1,85 @@
+    this.expansionModel.deselect(...toBeProcessed);
+  }
+
+  /** Gets a list of descendent data nodes of a subtree rooted at given data node recursively. */
```
This is subjective, but it'd make sense to me for abstract methods to be nearer to the top to make it clear what needs to be defined when I extend this class.

andrewseguin commented on src/cdk/tree/control/flat-tree-control.spec.ts:
```diff
@@ -0,0 +1,183 @@
+  describe('base tree control actions', () => {
+    it('should be able to expand and collapse dataNodes', () => {
+      const nodes = generateData(10, 4);
+      const node = nodes[1];
```
rename to `secondNode` to be consistent with `sixthNode`

andrewseguin commented on src/cdk/tree/control/flat-tree-control.spec.ts:
```diff
@@ -0,0 +1,183 @@
+
+      treeControl.toggle(sixthNode);
+
+      expect(treeControl.isExpanded(node)).toBeTruthy('Expect second node to stay expanded');
```
Add test to check that `isExpanded` is true for the sixthNode?

`expect(treeControl.isExpanded(sixthNode)).toBeTruthy('Expect sixth node to be expanded');`

andrewseguin commented on src/cdk/tree/control/flat-tree-control.ts:
```diff
@@ -0,0 +1,47 @@
+
+import {BaseTreeControl} from './base-tree-control';
+
+/** Flat tree control. Able to expand/collapse a subtree recursively for FlatNode type. */
```
What is `FlatNode type`?

tinayuangao replied:
`FlatNode` type is deleted. Updated to the doc.

andrewseguin commented on src/cdk/tree/control/flat-tree-control.ts:
```diff
@@ -0,0 +1,47 @@
+  }
+
+  /**
+   * Gets a list of descendent data nodes of a subtree rooted at given data node recursively.
```
"Gets a list of the data node's subtree of descendent data nodes" (seems like its a given that its recursive?). What does it mean that the data nodes must be "set correctly"?

tinayuangao replied:
Must be flattened tree nodes

andrewseguin commented on src/cdk/tree/control/flat-tree-control.ts:
```diff
@@ -0,0 +1,47 @@
+    const results: T[] = [];
+    let i = startIndex + 1;
+
+    for (;
```
Why is the for loop's initialization outside of the definition (`let i = startIndex + 1`)?

andrewseguin commented on src/cdk/tree/control/flat-tree-control.ts:
```diff
@@ -0,0 +1,47 @@
+    let i = startIndex + 1;
+
+    for (;
+        i < this.dataNodes.length && this.getLevel(dataNode) < this.getLevel(this.dataNodes[i]);
```
Is it possible to make this a function that can be documented? Would be nice to isolate the idea of this check and explain what it does

andrewseguin commented on src/cdk/tree/node.ts:
```diff
@@ -0,0 +1,127 @@
+   * For every node, there must be at least one when function that passes or an undefined to
+   * default.
+   */
+  when: (nodeData: T, index: number) => boolean;
```
Switch the params to match table and ngFor's trackby

andrewseguin commented on src/cdk/tree/tree-errors.ts:
```diff
@@ -0,0 +1,39 @@
+ * @docs-private
+ */
+export function getTreeMultipleDefaultNodeDefsError() {
+  return Error(`cdk-tree: There can only be one default row without a when predicate function.`);
```
Heads up that I recently removed the prefix `cdk-table` from the errors since I realized it was clear that it was being thrown from the table and it didn't make sense for users of `mat-table` to see `cdk-table`

andrewseguin commented on src/cdk/tree/tree.ts:
```diff
@@ -0,0 +1,222 @@
+              private _changeDetectorRef: ChangeDetectorRef) {}
+
+  ngOnInit() {
+
```
extra line here

tinayuangao commented on src/cdk/collections/tree-adapter.ts:
```diff
@@ -0,0 +1,28 @@
+  expandFlattenedNodes(nodes: T[], expansionModel: SelectionModel<T>): T[];
+
+  /**
+   * Put node descendants of node in array.
```
Deleted this part

tinayuangao commented:
Comments addressed. Please take another look. Thanks!

timeurg commented on src/cdk/tree/control/flat-tree-control.spec.ts:
```diff
@@ -0,0 +1,188 @@
+      expect(treeControl.expansionModel.selected.length)
+          .toBe(2, 'Expect two dataNodes in expansionModel');
+
+      treeControl.collapse(seconNode);
```
typo seconNode -> secondNode
